pkg/trace/sampler/scoresampler.go

```diff
@@ -89,3 +81,25 @@ func (s ScoreSampler) applySampleRate(root *pb.Span, rate float64) bool {
 	}
 	return sampled
 }
+
+// shrink limits the number of signatures stored in the sampler.
+// After a cardinality above shrinkCardinality/2 is reached
+// signatures are spread uniformly on a fixed set of values.
+// This ensures that ScoreSamplers are memory capped.
+// When the shrink is triggered, previously active signatures
+// stay unaffected.
+// New signatures may share the same TPS computation.
+func (s ScoreSampler) shrink(sig Signature) Signature {
+	if s.size() < shrinkCardinality/2 {
+		s.shrinkAllowList = nil
+		return sig
+	}
+	if s.shrinkAllowList == nil {
+		rates, _ := s.getAllSignatureSampleRates()
+		s.shrinkAllowList = rates
+	}
+	if _, ok := s.shrinkAllowList[sig]; ok {
+		return sig
+	}
+	return sig % (shrinkCardinality / 2)
+}
```

Comment:
There's something I'm not getting about this function.  Assignments to `shrinkAllowList` will not survive this function, because `ScoreSampler` is taken by value.  I don't see `shrinkAllowList` referenced elsewhere. So, the function is the same as:
```go
func (s ScoreSampler) shrink(sig Signature) Signature {
	if s.size() < shrinkCardinality/2 {
		return sig
	}

	rates, _ := s.getAllSignatureSampleRates()
	if _, ok := rates[sig]; ok {
		return sig
	}

	return sig % (shrinkCardinality / 2)
}
```
But in that case, it could instead be a method of `*Sampler`.

Haven't slept much, could be wrong.

Reply:
Thanks, updated. I'm keeping the shrink at the ScoreSampler level though as it's specific for Error and NoPriority samplers which have high cardinality of signatures